Site Rework

Erag0n001 opened the pull request:
### Short and concise description about my pull request:

- Reworked sites to be easier to work with, both for the players and server owners.

### TODOs:

- [X] - I confirm this PR has been previously tested by me and has no apparent issues.
- [X] - I confirm this PR is complying with this project's [Contribution Guidelines](https://github.com/RimworldTogether/Rimworld-Together/blob/development/.github/CONTRIBUTING.md).
- [X] - I confirm this PR is complying with this project's [Syntax Ruleset](https://github.com/RimworldTogether/Rimworld-Together/blob/development/.github/SYNTAX.md).

### Contributing:

- [X] - This PR implements a new usable feature or modifies an existing one considerably.
      
If the above is true, I have also uploaded a PR to this project's [Guide](https://github.com/RimworldTogether/Guide) so users can find information about it.

The link to said PR is: https://github.com/RimworldTogether/Guide/pull/13

### Longer / More informative description about what my pull request does:

- Faction sites are gone
- Personal sites no longer require pawns. This means players need to evacuate their pawns before the update happens. Recommend to announce it beforehand
- Personal sites now benefit the entire faction
- Personal sites now follow the player's faction, even after he leaves.
- Sites now have far more customizable rewards, set by the server owner.
- New Reward system where the server can give a def and amount to the client.
- Update command
- Now includes a good set of default values.
- Sites can now be configured by each client to accept from a list of reward instead of being forced to pick one.
- Reworked site UI.
- Added new site "Hunting Camp"

Erag0n001 committed:
Added client side UI for sites

Added client side UI for sites
- Data still needs to be set manually.

Erag0n001 committed:
Removed duplicate description

Erag0n001 committed:
Merge branch 'development' into SiteImplementation

Erag0n001 committed:
Reworked sites, still WIP

- Personal sites were removed
- Sites now send rewards
- New UI for sites
- Added rewards, which is the server telling a def and an amount to give to a client.

Erag0n001 committed:
Fixed factions and rewards

- Owners will now properly keep their sites when leaving, deleting and creating factions
- Rewards now properly arrive at their destination.

Erag0n001 committed:
Sites now properly sync with new config

Erag0n001 committed:
Fixed build error with Shared

Erag0n001 committed:
Fixed bugs and added backward compatibility.

- New update command to update from old sites to new sites.
- Fixed issue where good will wouldn't update properly when leaving and getting kicked from factions.
- Cleaned up some code
- Added descriptions to sites.

Erag0n001 committed:
Added client side UI to allow picking between rewards

Erag0n001 committed:
Merge branch 'development' into SiteImplementation

Erag0n001 committed:
Server side implementation of custom rewards

- Server now has "balanced" default values
- Server now remembers and uses the user's site config
- New client side button for config menu

Erag0n001 committed:
Fixed formatting error

Erag0n001 committed:
Fixed a few null checks and Shared

Erag0n001 committed:
Final fixes

- Fixed bug where rewards would be sometime doubled.
- Fixed bug where sites didn't update their factionfiles properly.
- Removed redundant variable in Sites.
- Fixed typos in config.
- Added a new site (it was for testing, but I already did the work so it stays)
- Added a few verbose logs for troubleshooting sake

Erag0n001 committed:
Fix critical bug

- Fixed where faction members could destroy sites.

Erag0n001 committed:
Removed redundant check

Byte-Nova reviewed (changes requested):
PR isn't able to compile correctly right now on Linux, it will be put into low priority until it gets fixed or someone is able to make it work correctly

Erag0n001 committed:
Fixed issue with building with linux

Issue came from windows being a dumb with case sensitivity. God I hate windows

Byte-Nova reviewed (changes requested):
Implementation seems to work as expected. I'd like to add a couple recommendations and considerations to it before it gets further tested and pushed.

**Consider specifying logs when values are null**
No warning or error is specified when thing to receive can't be found (def has been incorrectly set or is missing due to DLC item). Great job on catching a possible error and setting a default value when this one is missing out of the box.

``Example case:``
Site is configured to allow players to receive a DLC item, player that doesn't know what the item is checks it as a reward out of curiosity, doesn't receive anything nor any warning is displayed regarding the case.

**Reference values instead of setting them as one**
Site rewards are set as string when setting preferences, meaning that they don't react correctly when server admins change the rewards that a site can give.

``Example case:``
Server owner accidentally puts OP item in a site as a reward, notices and changes it, but players with the site reward already selected don't get it changed because it's a string, not a reference to something like the site reward list.

Erag0n001 committed:
Added requested changes

- Rewards will now prompt an error when failing to find a def.
- Sites will now prevent opening the UI if the cost or rewards are non existent
- Sites will now exclude any non-existing rewards or cost defs.
- Configs now sync properly on server restarts

Erag0n001 committed:
Merge branch 'development' into SiteImplementation

Erag0n001 committed:
Merged with dev and fixed bugs.

- Implemented new packet system
- Fixed reward managers having issues.
- Fixed issue with configs not being applied properly.

Erag0n001 committed:
Removed debugging log

Byte-Nova committed:
Merge branch 'development' into pr/167

Byte-Nova committed:
Fixed typos and changed textures for consistency

Byte-Nova reviewed (changes requested):
Found a couple of issues while testing the latest push of this PR:

**Players can't set their site configs:**
Site configurations are broken in this push, when trying to save them, the server throws an exception and kicks the client. Tried with a clean slate server and got the same results.

**Some sites can't be configured correctly because they have specific rewards:**
Depending on the name or the reward (specifically the letters composing it), the current scroll view will fail to render them correctly on the screen, ocluding them partially in some cases. Check the ´´Quarry´´ for an example or overload any site with different things to see it in greater detail.

Erag0n001 committed:
Fixed issue when users tried setting configs for the first time

Erag0n001 committed:
Fixed bug where text would sometime get out of bounds in site UI

Byte-Nova commented:
Moving into testing branch for further tests
